### PR TITLE
[PRP-347] Participant Documentation

### PR DESCRIPTION
**Pull request by microwavenby**
## Ticket

https://wicmtdp.atlassian.net/browse/PRP-347

* Documentation pages in `/docs/participant`
* New documentation pages rendered in storybook
* Breadcrumbed form page flow in Storybook

### CODE REVIEW

File: participant/.storybook/main.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ module.exports = {
     builder: "webpack5",
     disableTelemetry: true,
   },
-  staticDirs: ["../public"],
+  staticDirs: ["../public", { from: "../app", to: "participant/app" }],
```
Comment (rocketnova) [inline participant/.storybook/main.js:17]: 🤯 Whoa! I didn't know you could do that.

File: docs/participant/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+## Architecture and Overview
+
+The Participant portal is controlled by a [Remix](https://remix.run) application written in [Typescript](https://www.typescriptlang.org/).
+
+Form data submitted by participants is stored in a [Postgres](https://www.postgresql.org/) relational database, accessed through the [Prisma ORM library](https://www.prisma.io/). This database is run as a [Docker](docker.com/) container locally, and as a managed [Aurora Serverless V2 database](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-serverless-v2.html) in deployment.
+
+Uploaded documents submitted by participants are stored in a [Minio](https://min.io/) docker container locally, or in [AWS S3](https://docs.aws.amazon.com/s3/index.html) when deployed.
+
+![Participant Diagram](https://github.com/navapbc/wic-participant-recertification-portal/assets/723391/c1339576-38b8-4a92-b090-b7c52940b04a)
```
Comment (rocketnova) [inline docs/participant/README.md]: I find the first diagram slightly confusing because it looks like the psql container is not also run by docker. Ideas for resolving include, but are not limited to:
- Making remix, psql, and minio stacked vertically with arrows pointing to docker 
- OR removing the docker icon as a step and making it a box around remix, psql, and minio
- OR something else?

File: docs/participant/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+## Components
+
+### Remix (Participant Application)
+
+Remix is a javascript framework based on [React Router](https://reactrouter.com/en/main) and [React](https://react.dev/). It provides semantics for writing code that renders both a page for a browser and the HTTP API endpoints that page may consume, with "Loaders" responding to GET requests from the page, and "Actions" responding to POST requests. These semantics mean that one page such as the [count](../../participant/app/routes/%24localAgency/recertify/count.tsx) route can have a Loader to provide data needed for the page render, and an Action to process data submitted from that page's form.
```
Comment (rocketnova) [inline docs/participant/README.md]:
```suggestion
Remix is a javascript framework based on [React Router](https://reactrouter.com/en/main) and [React](https://react.dev/). It provides semantics for writing code that renders both a page for a browser and the HTTP API endpoints that the page may consume, with ["Loaders"](https://remix.run/docs/en/1.16.1/route/loader) responding to GET requests from the page, and ["Actions"](https://remix.run/docs/en/1.16.1/route/action) responding to POST requests. These semantics mean that one page, such as the [count](../../participant/app/routes/%24localAgency/recertify/count.tsx) route, can have a Loader to provide data needed for the page render, as well as an Action to process data submitted from that page's form.
```

File: docs/participant/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+The route's render occurs mostly with familiar [React](https://react.dev/) component patterns, and additional hooks to provide data from the Loader and Action functions on the server to the page.
+
+Remix cannot be run as a Single-Page Application, and must have a running server to host the page. As all data from an individual participant needs to be stored on a server, we consider this a reasonable limitation for the use case.
```
Comment (rocketnova) [inline docs/participant/README.md]:
```suggestion
Remix cannot be run as a Single-Page Application, and must have a running server to host the page. In the participant application, all data from an individual participant needs to be stored on a server, so we consider this a reasonable limitation for the use case.
```

File: participant/stories/docs/Readme.stories.mdx [hunk 5]
```diff
@@ -1,12 +1,14 @@
...
-## This is an import of the project's Readme for convenience
+# 📝 Overview
 
 ---
 
 <Description>{Readme}</Description>
+
+# Next Steps
+
+- How to develop on the [Participant Portal](?path=/story/docs-%F0%9F%9B%A0%EF%B8%8F-development--page)
```
Comment (rocketnova) [inline participant/stories/docs/Readme.stories.mdx]:
```suggestion
- How to [develop](?path=/story/docs-%F0%9F%9B%A0%EF%B8%8F-development--page) on the Participant Portal
```

File: participant/stories/docs/Development.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import Development from "../../../docs/participant/development.md";
+import { Meta, Description } from "@storybook/addon-docs";
+
+<Meta title="Docs/🛠️ Development" />
+
+# 🛠️
```
Comment (rocketnova) [inline participant/stories/docs/Development.stories.mdx:6]: This should either have a title or the README should remove the extra `# Overview` title

File: docs/participant/development.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
+# Development
+
+This application can be run directly on a developer's computer, or in a parallel to the deployment environments it can be run in [Docker](docker.com/). The object storage and relational database needs could be served by any compatible solution, but by default are run using [Docker Compose](https://docs.docker.com/compose/) in containers.
```
Comment (rocketnova) [inline docs/participant/development.md:3]:
```suggestion
This application can be run directly on a developer's computer, or in a parallel to the deployment environments, it can be run locally in [Docker](https://docker.com/). The object storage and relational database needs could be served by any compatible solution, but by default are run using [Docker Compose](https://docs.docker.com/compose/) in containers.
```

File: docs/participant/development.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+You can install it [directly by following the instructions](https://github.com/nvm-sh/nvm) on the Github repository
+
+You could also utilize [Brew](https://brew.sh/) on MacOS by running `brew install nvm` on the command line (after installing Brew)
+
+#### Using NVM to install NodeJS
+
+1.  Open your command line utility
+2.  Change directories to the `participant` folder `cd participant`
+3.  Install Node using `nvm install v19.1.0`
+4.  Instruct NVM to `nvm use v19.1.0`
```
Comment (rocketnova) [inline docs/participant/development.md:36]: We're using node 18, so these directions should be changed to reflect that.

```suggestion
2.  Change directories to the `participant` folder: `cd participant`
3.  Install Node: `nvm install v19.1.0`
4.  Instruct NVM to use the installed version: `nvm use v19.1.0`
```

File: docs/participant/development.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+#### Run NPM install to install packages
+
+Now that you have NVM to manage Node versions, and a compatible version of NodeJS, you can
+install the necessary packages for the Participant application using `npm`
+
+This should only need to happen when the contents of `package.json` change
+
+1.  Open your command line utility
+2.  Change directories to the `participant` folder with `cd participant`
+3.  Run `npm install -D` (this will take several minutes)
```
Comment (rocketnova) [inline docs/participant/development.md:47]:
```suggestion
1.  Open your command line utility
2.  Change directories to the `participant` folder: `cd participant`
3.  Install packages (this will take several minutes): `npm install -D`
```

File: docs/participant/development.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+This command copies the example environment file so that the application code can read these variables and connect to
+the Docker hosted services.
+
+1.  `cp .env.example .env`
+
+This command does several things - starts the postgres container, migrates the database schema, inserts seed records for clinic sites, builds the CSS stylesheets, starts the
+Minio S3 local container, creates an S3 Bucket to store files in, and finally starts the Remix service.
+
+2.  `npm run dev`
```
Comment (rocketnova) [inline docs/participant/development.md:67]: The formatting for this is hard to follow. Suggested edit:

```suggestion
This command copies the example environment file so that the application code can read these variables and connect to
the Docker hosted services.

1.  Copy the example environment variables file so that the application code can read these variables and connect to the Docker hosted services: `cp .env.example .env`
2.  Start the server: `npm run dev`
  - This command does several things. It starts the postgres container, migrates the database schema, inserts seed records for clinic sites, builds the CSS stylesheets, starts the Minio S3 local container, creates an S3 Bucket to store files in, and finally starts the Remix service.
```

File: docs/participant/development.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+From the `/participant` directory:
+
+`docker compose up`
+
+### 💻📦 About the local version of the application
+
+Changes to the code in the `participant` folder will cause the app to dynamically restart automatically.
+
+Changes to CSS need to be built, using the `npm run css` command.
```
Comment (rocketnova) [inline docs/participant/development.md:91]:
```suggestion
Changes to CSS need to be manually triggered: `docker compose exec remix_dev npm run css`
```

File: docs/participant/development.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+Now that you have NVM to manage Node versions, and a compatible version of NodeJS, you can
+install the necessary packages for the Participant application using `npm`
+
+This should only need to happen when the contents of `package.json` change
+
+1.  Open your command line utility
+2.  Change directories to the `participant` folder with `cd participant`
+3.  Run `npm install -D` (this will take several minutes)
+
+### 💻 Run the Remix application outside of Docker
```
Comment (rocketnova) [inline docs/participant/development.md:49]: I think you can decrease the headings by one level from here onwards so that "Prerequisites", Running outside of Docker, Running inside of docker, Form submissions, and Document submissions are all at the same level

File: docs/participant/development.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+`docker compose up`
+
+### 💻📦 About the local version of the application
+
+Changes to the code in the `participant` folder will cause the app to dynamically restart automatically.
+
+Changes to CSS need to be built, using the `npm run css` command.
+
+#### Form Submissions
```
Comment (rocketnova) [inline docs/participant/development.md:93]: This shouldn't be nested under Running the app in Docker

File: docs/participant/development.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+Form submissions will be stored in the local Docker postgres database. This should be durable across restarts of the application
+stack.
+
+You can connect to this database to view or alter form submissions with a postgres-compatible database client.
+
+With Brew on a Mac, you can:
+
+1.  Install the command line client with `brew install libpq`
+2.  Connect to the database with `psql -h127.0.0.1 -Upostgres`
```
Comment (rocketnova) [inline docs/participant/development.md:103]: Not sure if you want to include this, but you can also avoid the additional brew install and run `docker compose exec postgres psql -U postgres`

File: docs/participant/pages/00_index.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+The Index page is the landing page for the application.
+
+### About this page:
+
+This page contains static content about who is eligible to use this recertification portal.
+
+### Considerations for other states:
+
+We predict state agencies may make content and logo changes on this page
+Also, state agencies would need to remove the portal survey, or replace it with a new survey
```
Comment (rocketnova) [inline docs/participant/pages/00_index.md:12]: Some of these suggestions (header and footer) may not be right for this page, so feel free to move them to another page if that seems better.

```suggestion
- Replace the name of the site in the header to refer to your WIC state agency
- Replace the logos and content in the footer to refer to your WIC state agency
- Change the content on this page to refer to your WIC state agency
- Remove or replace the feedback survey link
```

File: docs/participant/pages/00_index.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
+# Index Page
+
+The Index page is the landing page for the application.
+
+### About this page:
+
+This page contains static content about who is eligible to use this recertification portal.
+
+### Considerations for other states:
```
Comment (rocketnova) [inline docs/participant/pages/00_index.md:9]: Let's rename this something that doesn't require the user to remember the project's historical context. Maybe "Considerations for a new deployment:" or "Considerations for a WIC state agency:" or "Considerations for your deployment"?

Whatever it ends up being, please make it consistent across all the subsequent pages.

File: docs/participant/pages/00_index.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+### Development
+
+Remix allows us to use URL parameters in our paths, so the file on our filesystem is [app/routes/$localAgency/recertify/index.tsx](../../../participant/app/routes/%24localAgency/recertify/index.tsx)
+
+The `$localAgency` parameter associates the participant's submission with a `LocalAgency` database record, using its `urlId` property.
+
+Invalid `urlId` values will be redirected to the first database record in the `LocalAgency` table, and an empty `LocalAgency` table will result in an error.
+
+The contents of this page are controlled by two things -
```
Comment (rocketnova) [inline docs/participant/pages/00_index.md:22]: Please make this change across all the subsequent pages.

```suggestion
The contents of this page are controlled by:
```

File: docs/participant/pages/03_count.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
+# Count
+
+### About these pages:
+
+User research showed that WIC participants don't distinguish between household members directly enrolled in WIC and authorized representatives. However, we only want to collect detailed information about WIC participants who are recertifying. So the count page includes content and nudges to help the user understand whose info we need. We use the /count page to identify how many people are recertifying from their household, which creates that number of "participant cards" on the /details page. The number entered in /count is not saved anywhere, it's just to generate participant cards. Users can remove the number of participant cards on the /details page.
```
Comment (rocketnova) [inline docs/participant/pages/03_count.md:5]: > The number entered in /count is not saved anywhere, it's just to generate participant cards. 

I think this is not true. I think this is saved to the db.

File: docs/participant/pages/03_count.md [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+### Considerations for other states:
+
+If your state is able to surface who is recertifying, for example by using MIS data, then you can avoid this and the /count pages all together and instead surface to the user who is recertifying.
```
Comment (rocketnova) [inline docs/participant/pages/03_count.md:9]: This is a bit confusing because:
1. There are generally no MIS APIs
2. There's no current way for the Remix app to receive this information or associate it with the user who is in the recertification flow

For simplicity, I suggest dropping this.

File: docs/participant/pages/04_details.md [hunk 20]
```diff
@@ -0,0 +1,21 @@
+# Details
+
+### About this page:
+
+Users fill out "participant cards" (a repeatable fieldset) to collect detailed information about participants who are recertifying. This info will both help staff find the participant in M-SPIRIT and ensure we request the right kinds of documents. Because WIC participants do not know their household ID and using EBT # would require additional steps for staff, we will rely on the participant's name and date of birth as the data points staff can use to look up the participant in M-SPIRIT. We ask adjunctive eligibility per participant since it's possible one participant has it and another doesn't.
```
Comment (rocketnova) [inline docs/participant/pages/04_details.md:5]: I think we should make this more generic. The audience is not necessarily a state using M-SPIRIT

```suggestion
Users fill out "participant cards" (a repeatable fieldset) to collect detailed information about participants who are recertifying. This info will both help staff find the participant in the MIS and ensure we request the right kinds of documents. Because WIC participants do not know their household ID and asking for their EBT # would require additional steps for staff, we rely on the participant's name and date of birth as the data points staff use to look up the participant in the MIS. We ask adjunctive eligibility per participant since it's possible one participant is adjunctively eligible and another is not.
```

File: docs/participant/pages/04_details.md [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+### Considerations for other states:
+
+If your state is able to surface who is recertifying, for example by using MIS data, then you can avoid this and the /count pages all together and instead surface to the user who is recertifying.
+
+### Development
+
+The Details page renders a number of [`<ParticipantCard>`](../../../participant/app/components/ParticipantCard.tsx) components, initially based on the `?count` parameter in the URL.
```
Comment (rocketnova) [inline docs/participant/pages/04_details.md:13]:
```suggestion
The Details page renders a number of [`<ParticipantCard>`](../../../participant/app/components/ParticipantCard.tsx) components, initially based on the `?count` query parameter in the URL.
```

File: docs/participant/pages/05_changes.md [hunk 22]
```diff
@@ -0,0 +1,21 @@
+# Changes
+
+### About this page:
+
+This question allows us to understand if something has changed with the WIC participant that would require them to upload new proof of address/identity docs. Montana WIC only requires documentation if these two scenarios have changed, or if any participant lacks adjunctive eligibility.
```
Comment (rocketnova) [inline docs/participant/pages/05_changes.md:5]:
```suggestion
This question allows us to understand if something has changed with the WIC participant that would require them to upload new proof of address/identity docs. Montana WIC, the WIC state agency for whom this application was originally developed, only requires documentation for recertification if these two scenarios have changed, or if any participant lacks adjunctive eligibility.
```

File: docs/participant/pages/05_changes.md [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+### Considerations for other states:
+
+When do you require recertifying WIC participants to upload documents? If only certain scenarios, what questions can you replace with these to understand those participants?
+For states that always require documents during recertification, are these questions helpful for local staff? Or, can you remove these questions and logic altogether?
```
Comment (rocketnova) [inline docs/participant/pages/05_changes.md:10]:
```suggestion
- Under what circumstances does your WIC state agency require WIC participants to provide documents during recertification appointments? If only certain scenarios, consider replacing these questions with ones that better fit those scenarios.
- If your WIC state agency always requires documentation during recertification, would the answers to these questions help WIC staff prepare for recertification appointments? If not, consider removing this page and bypassing the associated business logic altogether.
```

File: docs/participant/pages/06_upload.md [hunk 24]
```diff
@@ -0,0 +1,28 @@
+# Upload
+
+### About this page:
+
+​​We use the participant's situation to determine which kinds of documents, if any, they need to provide:
```
Comment (rocketnova) [inline docs/participant/pages/06_upload.md:5]:
```suggestion
Based on the answers the participant provided to previous questions, we determine which kinds of documents, if any, they need to provide:
```

File: docs/participant/pages/06_upload.md [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+- If they don't have adjunctive eligibility, they always need to upload proof of income.
+- If they moved, they need to provide proof of address.
+- If they changed their name and/or their previous ID doc expired, they need to provide new proof of identity.
+
+### Considerations for other states:
+
+What are your policies around verifying identity and address?
+If you require documents in all scenarios, you may be able to simplify this document upload flow.
+
+### Development
```
Comment (rocketnova) [inline docs/participant/pages/06_upload.md:16]: It could be worth describing the "previously uploaded" behaviour that happens on this page.

File: docs/participant/pages/07_contact.md [hunk 26]
```diff
@@ -0,0 +1,22 @@
+# Contact
+
+### About this page:
+
+We want staff to have a contact method in case there's issues with the submission.
+We also want to nudge participants to share any nutrition or life updates, but testing with Montana WIC recipients revealed participants might not always have an update to share.
+
+### Considerations for other states:
+
+Is there anything else you need to collect ahead of a recertification appointment?
```
Comment (rocketnova) [inline docs/participant/pages/07_contact.md]:
```suggestion
Is there anything else you need to collect ahead of a recertification appointment? It's important to collect all necessary information, but be sure to balance that against collecting too much information. Asking participants to fill out excessively long forms or asking for sensitive PII may make participants reluctant to use the application.
```

File: docs/participant/pages/09_confirm.md [hunk 27]
```diff
@@ -0,0 +1,25 @@
+# Confirm
+
+### About this page:
+
+This page confirms the submission and shows content about next steps. It contains an optional user survey for feedback about their experience. It shows the user's submitted information.
+
+### Considerations for states:
+
+- You may want to update next steps content with agency specific next steps.
+- You can replace your own survey link with your own to collect user feedback.
```
Comment (rocketnova) [inline docs/participant/pages/09_confirm.md:10]:
```suggestion
- Replace the survey link with your own to collect user feedback.
```

File: docs/participant/pages/09_confirm.md [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+
+### Development
+
+The Confirm page renders all previous page data stored in `SubmissionForm` (and a count of `Document` records) in
+a special component (not in Storybook) called a [`<SubmissionForm>`](../../../participant/app/components/SubmissionForm.tsx)
+
+The applicant cannot edit this page, and navigation will ask the individual if they would like to reset their cookie
+to allow completing the form a second (or nth) time.
```
Comment (rocketnova) [inline docs/participant/pages/09_confirm.md:18]:
```suggestion
The Confirm page renders all previous page data stored in `SubmissionForm` (and a count of `Document` records) in a special component (not in Storybook) called a [`<SubmissionForm>`](../../../participant/app/components/SubmissionForm.tsx)

Once the participant reaches this page, they are not longer able to edit any information. If they click the back button in their browser, an alert will appear asking the individual if they would like to reset their cookie to allow completing the form a second (or nth) time.
```

File: docs/participant/pages/04_details.md [hunk 29]
```diff
@@ -0,0 +1,21 @@
+# Details
+
+### About this page:
+
+Users fill out "participant cards" (a repeatable fieldset) to collect detailed information about participants who are recertifying. This info will both help staff find the participant in M-SPIRIT and ensure we request the right kinds of documents. Because WIC participants do not know their household ID and using EBT # would require additional steps for staff, we will rely on the participant's name and date of birth as the data points staff can use to look up the participant in M-SPIRIT. We ask adjunctive eligibility per participant since it's possible one participant has it and another doesn't.
+
+### Considerations for other states:
+
+If your state is able to surface who is recertifying, for example by using MIS data, then you can avoid this and the /count pages all together and instead surface to the user who is recertifying.
```
Comment (rocketnova) [inline docs/participant/pages/04_details.md:9]: Same comment as for count. Suggesting dropping this.

```suggestion
If your state is able to surface who is recertifying, for example by using MIS data, then you can avoid this and the /count pages all together and instead surface to the user who is recertifying.
```

Please add something like this:
```
- Replace the adjunctive eligibility question with the programs that qualify a participant for WIC adjunctive eligibility in your state
```

**Review by rocketnova — COMMENTED** [review]
This is looking great! I've gone through and made a number of edits and suggestions. In addition, could you add documentation for these items:

- Document how local agencies are created in the database and how to replace those with the correct ones for "your WIC state agency"
- Document how to insert pages or remove pages from the routing structure
- How to change the session timeout

File: docs/participant/pages/06_upload.md (hunk 25 above)
Comment (microwavenby) [reply docs/participant/pages/06_upload.md:16]: Did it to 11 💥

**Comment by microwavenby** [comment]
> This is looking great! I've gone through and made a number of edits and suggestions. In addition, could you add documentation for these items:
> 
>     * Document how local agencies are created in the database and how to replace those with the correct ones for "your WIC state agency"
> 
>     * Document how to insert pages or remove pages from the routing structure
> 
>     * How to change the session timeout

Added new readme file `docs/participant/application.md` to talk about routing, database, internationalization, and session

**Review by rocketnova — APPROVED** [review]
